ls1intum/Artemis
# [Feature] Metis Emoji Reaction on Postings

### Checklist

### Motivation and Context


Introduction of Emoji related features for Metis.

### Description

This PR introduces a new reaction feature for Metis Postings (similar to slack). With this change, the old vote component (with arrows up and down) is replaced. Support/Agreement for a postings can now be expressed by a certain emoji, initially, this is not configurable and set to the '+' emoji.

This PR required DB migration, the emojiId column of the reaction table has to be changed to allow larger strings (some of them are really long). The visionary data model looks like this:

### Steps for Testing


1. Log in to Artemis 
2. Navigate to a course with posts enabled
3. Navigate to an exercise
4. React on existing posts and answer post, check the following:
i)  Selecting two times the same emoji will remove the emoji, either by selecting it twice via the selector or when clicking on the displayed little emoji icon again
ii) When selecting the '+', the order of the posts might change -> posts are sorted by the amount of '+' first, then by date (newest at the bottom, to have chronology from top-down)
5. Repeat step 4. for a lecture
6. As instructor, check if the discussion overview still works properly, as this is a related component -> the votes column should now display the amount of '+' reactions on a posting
7. As instructor/tutor/student create/edit/delete postings, approve (as tutor) as these are related components

### Test Coverage



Server:
PostResource: 100%
PostingService: 100%
PostService: 100%
AnswerPostResource: 100%
AnswerPostService: 100%
ReactionResource: 100%
ReactionService: 100%

Client:
post-reaction-bar.component.ts: 100%
answer-post-reaction-bar.component.ts: 100%
postings-reaction-bar.component.ts: 85%


### Screenshots



How a student can react on postings with new reactions and existing reactions:

https://user-images.githubusercontent.com/41366522/128930763-29d034b0-22cd-41c3-b684-29a618339e1f.mov
